### ISSUE TITLE
FBBT can have math domain error in log()

### ISSUE THREAD

**Issue by qtothec** [Pyomo/pyomo#969]
In interval,py, we have the following:
```
def log(xl, xu):
    if xl > 0:
        return math.log(xl), math.log(xu)
    elif xl == 0:
        if xu > 0:
            return -math.inf, math.log(xu)
        else:
            return -math.inf, -math.inf
    else:
        return -math.inf, math.inf
```

If `xu = 0`, then `math.log(xu)` raises a ValueError.

@michaelbynum did you mean to have:
```
def log(xl, xu):
    if xl > 0 and xu > 0:
        return math.log(xl), math.log(xu)
    elif xl == 0:
        if xu > 0:
            return -math.inf, math.log(xu)
        else:
            return -math.inf, -math.inf
    else:
        return -math.inf, math.inf
```

I'm also a bit confused about the branch that returns `-math.inf, -math.inf`.

**Comment by michaelbynum** [comment]
@qtothec I had assumed that ```xu >= xl```. If that assumption is true, then if ```xu=0```, then ```xl <= 0```. In that case, I don't think ```math.log``` is called. However, my assumption that ```xu >= xl``` may be a bad one. 

```-math.inf, -math.inf``` is only returned if ```xl = xu = 0```, which I think is correct.

**Comment by qtothec** [comment]
I tried turning on fbbt with GDPopt-local and at least one case in the current test set ended up violating the assumption. I can take a closer look at it later.

**Comment by qtothec** [comment]
Tangentially related to this issue, I run into float division by zero on a linear relaxation of this problem (technically, while solving the set covering master problem. Code from test_gdpopt):
```
    def test_infeasible_GDP(self):
        """Test for infeasible GDP."""
        m = ConcreteModel()
        m.x = Var(bounds=(0, 2))
        m.d = Disjunction(expr=[
            [m.x ** 2 >= 3, m.x >= 3],
            [m.x ** 2 <= -1, m.x <= -1]])
        m.o = Objective(expr=m.x)
        output = StringIO()
        with LoggingIntercept(output, 'pyomo.contrib.gdpopt', logging.WARNING):
            SolverFactory('gdpopt').solve(
                m, strategy='LOA',
                mip_solver=mip_solver,
                nlp_solver=nlp_solver)
            self.assertIn("Set covering problem was infeasible.",
                          output.getvalue().strip())
```
```
  File "/home/qichen/git/pyomo/pyomo/contrib/fbbt/fbbt.py", line 983, in fbbt
    integer_tol=integer_tol)
  File "/home/qichen/git/pyomo/pyomo/contrib/fbbt/fbbt.py", line 930, in fbbt_block
    integer_tol=integer_tol)
  File "/home/qichen/git/pyomo/pyomo/contrib/fbbt/fbbt.py", line 838, in fbbt_con
    visitorB.dfs_postorder_stack(con.body)
  File "/home/qichen/git/pyomo/pyomo/core/expr/visitor.py", line 518, in dfs_postorder_stack
    flag, value = self.visiting_potential_leaf(_sub)
  File "/home/qichen/git/pyomo/pyomo/contrib/fbbt/fbbt.py", line 759, in visiting_potential_leaf
    _prop_bnds_root_to_leaf_map[node.__class__](node, self.bnds_dict)
  File "/home/qichen/git/pyomo/pyomo/contrib/fbbt/fbbt.py", line 300, in _prop_bnds_root_to_leaf_ProductExpression
    _lb1, _ub1 = interval.div(lb0, ub0, lb2, ub2)
  File "/home/qichen/git/pyomo/pyomo/contrib/fbbt/interval.py", line 28, in div
    return mul(xl, xu, *inv(yl, yu))
  File "/home/qichen/git/pyomo/pyomo/contrib/fbbt/interval.py", line 24, in inv
    return 1.0/xu, 1.0/xl
ZeroDivisionError: float division by zero
```

**Comment by qtothec** [comment]
Related to the original `log()` issue,
I have this constraint:
```
rectangle_in_circle_disjuncts[3].constraint[2] (rect_x[2] - 3.5 - 50)**2 + (rect_y[2] - 2.5 - 80)**2
VarName LB UB:
rect_x[2] 48.60102051443364 51.5
rect_y[2] 72.91742430504416 76.5
```

```
  File "/home/qichen/git/pyomo/pyomo/contrib/fbbt/fbbt.py", line 985, in fbbt
    integer_tol=integer_tol)
  File "/home/qichen/git/pyomo/pyomo/contrib/fbbt/fbbt.py", line 915, in fbbt_block
    integer_tol=integer_tol)
  File "/home/qichen/git/pyomo/pyomo/contrib/fbbt/fbbt.py", line 840, in fbbt_con
    visitorB.dfs_postorder_stack(con.body)
  File "/home/qichen/git/pyomo/pyomo/core/expr/visitor.py", line 518, in dfs_postorder_stack
    flag, value = self.visiting_potential_leaf(_sub)
  File "/home/qichen/git/pyomo/pyomo/contrib/fbbt/fbbt.py", line 759, in visiting_potential_leaf
    _prop_bnds_root_to_leaf_map[node.__class__](node, self.bnds_dict)
  File "/home/qichen/git/pyomo/pyomo/contrib/fbbt/fbbt.py", line 396, in _prop_bnds_root_to_leaf_PowExpression
    _lb1, _ub1 = interval._inverse_power1(lb0, ub0, lb2, ub2, orig_xl=lb1, orig_xu=ub1)
  File "/home/qichen/git/pyomo/pyomo/contrib/fbbt/interval.py", line 145, in _inverse_power1
    xl, xu = log(zl, zu)
  File "/home/qichen/git/pyomo/pyomo/contrib/fbbt/interval.py", line 261, in log
    return math.log(xl), math.log(xu)
ValueError: math domain error
```

**Comment by michaelbynum** [comment]
@qtothec What are the values of ```rectangle_in_circle_disjuncts[3].constraint[2].lower``` and ```rectangle_in_circle_disjuncts[3].constraint[2].upper```?

**Comment by qtothec** [comment]
@michaelbynum 

> What are the values of `rectangle_in_circle_disjuncts[3].constraint[2].lower` and `rectangle_in_circle_disjuncts[3].constraint[2].upper`?

`None` and  `25.0`, respectively.

**Comment by michaelbynum** [comment]
@qtothec Regarding the math domain error for the ```log``` issue, the constraint you mentioned is infeasible, which is ultimately the problem. I should be able to detect this infeasibility before this error is hit so that a more useful error message is raised. I'll work on this.

**Comment by michaelbynum** [comment]
@qtothec Where can I find ```test_infeasible_GDP```? Is it on a branch on your fork?

**Comment by qtothec** [comment]
That's in the test_GDPopt.py file in master.

**Comment by michaelbynum** [comment]
@qtothec I am not getting the division by zero error you mentioned. Can you post the full stack?

**Comment by qtothec** [comment]
Full stack trace:
```
======================================================================
ERROR: Test for infeasible GDP.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/qichen/git/pyomo/pyomo/contrib/gdpopt/tests/test_gdpopt.py", line 166, in test_infeasible_GDP
    nlp_solver=nlp_solver)
  File "/home/qichen/git/pyomo/pyomo/contrib/gdpopt/GDPopt.py", line 343, in solve
    GDPopt_initialize_master(solve_data, config)
  File "/home/qichen/git/pyomo/pyomo/contrib/gdpopt/master_initialize.py", line 45, in GDPopt_initialize_master
    init_strategy_fctn(solve_data, config)
  File "/home/qichen/git/pyomo/pyomo/contrib/gdpopt/master_initialize.py", line 185, in init_set_covering
    set_cover_mip, disjunct_needs_cover, solve_data, config)
  File "/home/qichen/git/pyomo/pyomo/contrib/gdpopt/master_initialize.py", line 243, in solve_set_cover_mip
    mip_results = solve_linear_GDP(m.clone(), solve_data, config)
  File "/home/qichen/git/pyomo/pyomo/contrib/gdpopt/mip_solve.py", line 45, in solve_linear_GDP
    fbbt(m, integer_tol=config.integer_tolerance)
  File "/home/qichen/git/pyomo/pyomo/contrib/fbbt/fbbt.py", line 982, in fbbt
    integer_tol=integer_tol)
  File "/home/qichen/git/pyomo/pyomo/contrib/fbbt/fbbt.py", line 929, in fbbt_block
    integer_tol=integer_tol)
  File "/home/qichen/git/pyomo/pyomo/contrib/fbbt/fbbt.py", line 837, in fbbt_con
    visitorB.dfs_postorder_stack(con.body)
  File "/home/qichen/git/pyomo/pyomo/core/expr/visitor.py", line 518, in dfs_postorder_stack
    flag, value = self.visiting_potential_leaf(_sub)
  File "/home/qichen/git/pyomo/pyomo/contrib/fbbt/fbbt.py", line 759, in visiting_potential_leaf
    _prop_bnds_root_to_leaf_map[node.__class__](node, self.bnds_dict)
  File "/home/qichen/git/pyomo/pyomo/contrib/fbbt/fbbt.py", line 300, in _prop_bnds_root_to_leaf_ProductExpression
    _lb1, _ub1 = interval.div(lb0, ub0, lb2, ub2)
  File "/home/qichen/git/pyomo/pyomo/contrib/fbbt/interval.py", line 28, in div
    return mul(xl, xu, *inv(yl, yu))
  File "/home/qichen/git/pyomo/pyomo/contrib/fbbt/interval.py", line 24, in inv
    return 1.0/xu, 1.0/xl
ZeroDivisionError: float division by zero
```

The relevant constraint is:
```
_pyomo_gdp_bigm_relaxation.relaxedDisjuncts[0].d_disjuncts[0].constraint[2,lb]
body: x + 3.0*(1 - d_disjuncts[0].indicator_var)
lower: 3.0
upper: None

VarName LB UB:
x 0 2
d_disjuncts[0].indicator_var 1 0
```

**Comment by michaelbynum** [comment]
@qtothec I am having trouble reproducing this error. Can you either test my "fbbt_fix" branch to see if the problem is fixed or provide additional information (e.g., GDPOpt options) to help me reproduce the error?

**Comment by qtothec** [comment]
Yes, it looks like the `fbbt_fix` branch resolves this issue.

**Comment by michaelbynum** [comment]
Awesome!

**Comment by michaelbynum** [comment]
Addressed by PR #994
